kernel/cemitcodes.ml

```diff
@@ -369,10 +369,17 @@ let to_memory (init_code, fun_code, fv) =
   init();
   emit init_code;
   emit fun_code;
-  let code = String.create !out_position in
-  String.unsafe_blit !out_buffer 0 code 0 !out_position;
   (** Later uses of this string are all purely functional *)
-  let code = CString.hcons code in
+  let code = Bytes.init !out_position (fun i ->
+      Bytes.unsafe_get !out_buffer i
+    ) in
+  (* XXX: We lose the hconsing optimization wrt the non-bytes version.
+
+     This really depends on whether we want emitcodes to be `string` or `bytes`.
+
+     The current choice is `bytes`, but `string` can be done using a
+     more careful patching. *)
```

Comment:
Sure, I did not meant that using `unsafe` is incorrect here, but certainly it has a cost in terms of semantics and maintenance. IMVHO, `unsafe` it is really _the last resort_.

Imagine that in the future `string` is really compiled to an immutable datatype such as JS `string`,  I'd rather clean the code up now and never have to worry again.

Coq is large enough that reasoning is hard, so I'd rather let type system do the reasoning for me, see for instance the problem in #441 with the innocent `we_are_parsing` imperative flags, etc...